### PR TITLE
add methods to PudlTabl so it can be serialized and de-serialized (v2)

### PR DESCRIPTION
**Pull request by arengel**
Apologies for the duplicate PR but was having issues pushing to PUDL. Now that it is sorted, this PR replaces #2228. It isn't super urgent but I'm guessing it's easier not to have to futz with GitHub Actions. I've also updated this text based on additional changes I needed to make so that `Datastore` can be serialized, an issue that didn't arise with local testing.

## Motivation

The purpose of this PR is to enable `PudlTabl` to be serialized using `pickle`. In RMI's work with PUDL, we've often found it useful to save a copy of a `PudlTabl` before testing or running our data processing. This can save substantial amounts of time without jeopardizing validity because essentially all of our code is downstream of PUDL. The proposed changes here are a refined and generalized version of the solution we have used for a while now. In addition to being a more elegant solution than us subclassing or wrapping `PudlTabl`, it bring this functionality to all users, and the proposed tests should ensure that it continues to work properly.

## The changes

1. Two new methods for `PudlTabl`. A description of the default versions of these methods can be found [here](https://docs.python.org/3/library/pickle.html#pickle-inst), an example of how they might be modified in special cases is discussed [here](https://docs.python.org/3/library/pickle.html#pickle-state).
   1. `__getstate__` this converts the `PudlTabl` state into one that can be serialized with `pickle`. This is required because `pickle` doesn't like the `lambda` in the `self._dfs` `defaultdict` or `sa.Engine` so they need to be converted into form that `pickle` can deal with.
   2. `__setstate__` this takes the state as it was saved above and puts the pieces back into `PudlTabl` so a restored `PudlTabl` behaves the same as it would have before it was serialized. One important detail is that ``__init__`` is not run when an  object is de-serialized, so under some circumstances, setup that happened there may also need to occur here.
2. The same new methods for `GoogleCloudStorageCache`. This is required because one of its attributes is a `google.cloud.storage.Client` which cannot be pickled because "Pickling client objects is explicitly not supported. Clients have non-trivial state that is local and unpickleable." My solution here is to add `__getstate__` / `__setstate__` methods to `GoogleCloudStorageCache` which allow a new instance of it to be recreated based on the arguments from which the original object was created. This now allows `Datastore`, or at least the one I test with, to be serialized using pickle.
3. A new test
    1. `test_pudltabl_pickle` - this tests that we can take a `PudlTabl` with at least one df and that it can be pickled and restored from a pickle.

## A few additional points

1. My approach to saving the state of the `pudl_engine` is to save the URL, and on restoring, using `sa.create_engine` to create a new `sa.Engine` from the saved URL. This works fine. But then I thought, that will produce a `PudlTabl` that won't work properly if it is moved from the user/machine on which it was created. To fix this I added a check that tries connecting to the engine created from the saved URL, if that fails, we create an engine from the URL in `pudl.workspace.setup.get_defaults`. While that is probably convenient, it might cause weird problems I haven't thought of.
2. Similar to the issue with `pudl_engine`, there is a potential issue with this approach to `GoogleCloudStorageCache` and `Datastore` more generally where a restored `PudlTabl` contains a non-functioning `Datastore` because on a different machine, the `LocalFileCache` would need to point to a different place, or because `GoogleCloudStorageCache` would  be missing the proper credentials, or something like that. This raises a few questions:

   1. Is restoring the `Datastore` important, the primary use case that I envision for serializing `PudlTabl` is to be able to quickly and easily store/restore a set of DataFrames. Recreating the caches, etc. that created the DataFrames is, at least from my point of view, a lower priority. If this is a priority than the following questions would need to be addressed in some form.
   2. Does it make sense to include in this process a way to check that the restored `Datastore` functions properly and if not, either recreate it from local defaults or just avoid creating the non-functioning parts?
   3. If `PudlTabl` portability is important, any recommendations on how to test that this kind of orphaned `Datastore` (and really `PudlTabl`) works properly within the test suite? I think it means creating a pickle with one set of settings, and then restoring it in a place where `pudl.workspace.setup.get_defaults()` will return different values from the ones used in create the pickle in the first place. At least I think that's what is required.

3. The test uses the fixture `fast_out_annual`. This works totally fine when running the whole test suite but creating that fixture just to test serializing `PudlTabl` is overkill. For now this seemed like the better approach than creating another `PudlTabl`, especially since I wasn't sure what DB to connect it to so that it would work in GHA/CI land. If more work is required, especially as relates to the point below, it might be worth creating a minimal `PudlTabl` fixture for testing this functionality. Here I would request guidance on how to set it up so it plays nicely.

## PR Checklist

- [x] Merge the most recent version of the branch you are merging into (probably `dev`).
- [x] All CI checks are passing. [Run tests locally to debug failures](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)
- [x] Make sure you've included good docstrings.
- [x] For major data coverage & analysis changes, [run data validation tests](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#data-validation)
- [x] Include unit tests for new functions and classes.
- [x] Defensive data quality/sanity checks in analyses & data processing functions.
- [x] Update the [release notes](https://catalystcoop-pudl.readthedocs.io/en/latest/release_notes.html) and reference reference the PR and related issues.
- [x] Do your own explanatory review of the PR to help the reviewer understand what's going on and identify issues preemptively.

### CODE REVIEW

**Comment by arengel** [comment]
Moving this branch to PUDL proper, the CI tests run, and they fail or at least they did. The reason, as I sort of expected, is that `Datastore` cannot be pickled, at least not when it contains a `GoogleCloudStorageCache`. Broadly speaking, there are two ways to resolve this:
1. **Write `__getstate__` / `__setstate__` for `Datastore`**, maybe for the underlying cache objects as well so that all the proper state is saved and restored. This option becomes more compelling if there is state in all these objects of non-trivial complexity that can be meaningfully stored. I think this is more or less the 'right' way to do it, so that each object only has to know how to store/restore itself.
2. **`PudlTabl` handles a few elements of `Datastore` state**, this is what I did and is more or less analogous to the way I dealt with `sa.Engine`. This makes sense if there isn't much state in `Datastore` that can be meaningfully stored and all the cache objects sort of have to be recreated from scratch anyway. The current implementation keeps track of whether the `PudlTabl` had a `Datastore` and if it used the zenodo sandbox or not. When restoring a `PudlTabl` that had a `Datastore`, a new one is created from local defaults (i.e. `Datastore(local_cache_path=pudl_settings["data_dir"],...`) with the relevant sandbox setting.

I went with (2) because I think the tricky things in `Datastore` probably need to be recreated anyway. And because the primary use case that I envision for serializing `PudlTabl` is to be able to quickly and easily store/restore a set of DataFrames. Recreating the caches, etc. that created the DataFrames is, at least from my point of view, a lower priority. So trying to figure out how to serialize the caches seemed of lesser importance. I also noticed this, which, at least for the `GoogleCloudStorageCache`, makes me think serializing it just won't work:


``` python
class Client:

    def __getstate__(self):
        """Explicitly state that clients are not pickleable."""
        raise PicklingError(
            "\n".join(
                [
                    "Pickling client objects is explicitly not supported.",
                    "Clients have non-trivial state that is local and unpickleable.",
                ]
            )
        )
```

Definitely open to (1) or saving more state in (2) if the current approach is missing important elements.

**Review by jdangerx — CHANGES_REQUESTED** [review]
Hi! Thanks for re-opening the PR! I think the `pre-commit.ci` error should be OK if you update to the latest `dev`.

Separately, do you mind writing another test so our coverage stays high? Looks like it's the exception handling code that's not getting covered - which, arguably, because it's kind of the cleverest code, is pretty important to hit in test.

Thirdly - in case nobody else has mentioned this, we *are* thinking about deprecating `PudlTabl` in favor of actually persisting these outputs in the main SQLite database, "soon." No concrete timeline on that yet, but useful context for someone who's contributing to `PudlTabl`!

File: test/integration/output_test.py
Comment (jdangerx) [inline test/integration/output_test.py:290]: Another alternative would be to change the scope of `fast_out_annual` back to `function` (the default) - then we don't have to worry about inter-test entanglement. I just ran the tests both ways and neither seems slower than the other - what do you think?

**Review by jdangerx — APPROVED** [review]
Looks good! One nit-pick comment that you are free to ignore, but I'll give you a chance to engage before clicking the merge button :)

I agree that squash + merge would be the best option here - unless you feel strongly about cleaning up your history via `rebase -i`.

File: test/integration/output_test.py
Comment (arengel) [reply test/integration/output_test.py]: Agreed, that is better.

**Pull request merged**
